BTC/implementation-of-the-bitcoin-system.md

```diff
@@ -0,0 +1,211 @@
+# L5 Implementation of the Bitcoin System
+
+## Ledger Design Model
+
+- Blockchain: A decentralized ledger technology used in various systems.
+- Bitcoin’s Ledger Model: Uses a **transaction-based** approach where account balances are not directly recorded, requiring the replay of transaction history to calculate balances. Each transaction must specify the source of the coins.
+- Ethereum’s Ledger Model: Adopts an **account-based** ledger system that explicitly records each account’s balance, similar to traditional banking experiences.
+
+## UTXO
+
+```plain
+B ----------> A (5 BTC)
+C ----------> A (3 BTC)
+
+A ----------> D (8 BTC)
+```
+
+- UTXO (Unspent Transaction Outputs): A data structure that prevents double spending in Bitcoin.
+- Transaction Example:
+  - B sends 5 BTC to A; C sends 3 BTC to A, giving A two UTXOs.
+  - A sends 8 BTC to D, consuming these two UTXOs.
+	- Miners check against previous blocks to ensure no double spending.
+- UTXO Set:
+	- Maintained in memory by full nodes for quick double spending detection.
+	- Contains elements like:
+	  - Transaction Identifier (Txid): Unique hash of the transaction.
+	  - Output Index: Identifies the specific output within a transaction.
+- Transaction Process:
+	- Transactions consume old UTXOs and create new ones.
+	- Example: After A’s transaction, a new UTXO A ----------> D (8 BTC) is created.
+- UTXO Set Storage:
+	- Despite gradual growth, the UTXO set fits within standard server memory.
+
+## Total inputs `=` total outputs
+
+- Transaction Inputs and Outputs:
+  - A transaction can have multiple inputs and outputs.
+	- Typically, the sum of inputs exceeds the sum of outputs; the difference is the transaction fee for miners.
+	- If inputs equal outputs, there is no transaction fee.
+- Incentives in the Bitcoin System:
+	- Miners receive block rewards and transaction fees as incentives.
+	- As block rewards decrease, transaction fees will become the primary incentive for miners.
+	- Block Reward Halving:
+	- The block reward is halved every 210,000 blocks.
+	- This halving occurs approximately every 4 years, calculated as $$\frac{210000 \times 10  \text{ minutes}}{60  \text{ minutes} \times 24 \times 365} \approx 4 \text{ years}.$$
+
+## Bitcoin block information
+
+Bitcoin Block (548,402)
```

Comment:
Is it a height? I think you should mark it as an example?

Reply:
done